Reduce ROM usage due to X509 info

pkolbus opened the pull request:
## Description
Allow projects to save some ROM by omitting mbedtls_x509_*_info() and strings only used by these functions. Resolves #2308 

## Status
**READY**

## Requires Backporting
NO - this is an enhancement

## Migrations
NO (I don't think so, but see question below)

## Additional comments
n/a

## Todos
- [x] Tests
- [x] Documentation
- [x] Changelog updated
- [x] Backported (n/a)


## Steps to test or reproduce

- In small: `scripts/config.pl unset MBEDTLS_X509_INFO; make; make test` (this is included in all.sh; test-ref-configs.pl also exercises several configs with both options unset).
- Compare code size with and without these options set

RonEld commented:
As mentioned in #2313 , this is pending CLA as well

pkolbus commented:
Thanks for the feedback, will rework.

Just for clarification- would you consider mbedtls_x509_crt_verify_info (my second patch) as falling under the same, single configuration option?

hanno-arm commented:
@pkolbus At first I thought it could give independent value for debugging to have just the verification failure info enabled, but on second thought, a constrained device would rather just export the error code code and have the user look up its meaning manually. Therefore, I'd suggest you also include the removal of the verification info into the scope of the new option.

@mpg What's your view on that?

mpg commented:
@hanno-arm I agree. We want to support very constrained devices, and we don't want a plethora of config option because combinatorial explosion hurts testing. So we should have only one config option, and it should remove as much as sensible.

pkolbus commented:
@mpg @hanno-arm, thanks for the great feedback- makes perfect sense. It also looks like mbedtls_debug_print_crt() is basically useless without MBEDTLS_X509_INFO, so I'm planning to guard that whole function.

Should have a rebased and reworked PR tonight (US time); at the moment my PC is working through all.sh.

pkolbus commented:
The rebase from ba777bf to 9f54fc3 addresses feedback from @hanno-arm 
The rebase from 9f54fc3 to 49955ff updates the test in all.sh for #2325

In my environment, all.sh now stops in "build_default_make_gcc_and_cxx: test: verify header list in cpp_dummy_build.cpp"; I'll look into that this weekend.

pkolbus commented:
Hi all, sorry for the delay. Finally found some time to rebase and retest. This should be ready to review again.

hanno-arm commented on include/mbedtls/oid.h:
```diff
@@ -588,6 +591,7 @@ int mbedtls_oid_get_md_hmac( const mbedtls_asn1_buf *oid, mbedtls_md_type_t *md_
  * \return         0 if successful, or MBEDTLS_ERR_OID_NOT_FOUND
  */
 int mbedtls_oid_get_extended_key_usage( const mbedtls_asn1_buf *oid, const char **desc );
+#endif
```
Minor: Please add `/* MBEDTLS_X509_INFO */` here.

hanno-arm commented on include/mbedtls/x509_crl.h:
```diff
@@ -149,6 +150,7 @@ int mbedtls_x509_crl_parse_file( mbedtls_x509_crl *chain, const char *path );
  */
 int mbedtls_x509_crl_info( char *buf, size_t size, const char *prefix,
                    const mbedtls_x509_crl *crl );
+#endif
```
Minor: Please add `/* MBEDTLS_X509_INFO */` here.

hanno-arm commented on include/mbedtls/x509_crt.h:
```diff
@@ -362,7 +363,9 @@ int mbedtls_x509_crt_parse_path( mbedtls_x509_crt *chain, const char *path );
 int mbedtls_x509_crt_info( char *buf, size_t size, const char *prefix,
                    const mbedtls_x509_crt *crt );
+#endif
 
```
Minor: These two `#endif ... #if` can be merged.

hanno-arm commented on include/mbedtls/x509_crt.h:
```diff
@@ -377,6 +380,7 @@ int mbedtls_x509_crt_info( char *buf, size_t size, const char *prefix,
  */
 int mbedtls_x509_crt_verify_info( char *buf, size_t size, const char *prefix,
                           uint32_t flags );
+#endif
```
Minor: Please add `/* MBEDTLS_X509_INFO */` here.

hanno-arm commented on include/mbedtls/x509_csr.h:
```diff
@@ -137,6 +138,7 @@ int mbedtls_x509_csr_parse_file( mbedtls_x509_csr *csr, const char *path );
  */
 int mbedtls_x509_csr_info( char *buf, size_t size, const char *prefix,
                    const mbedtls_x509_csr *csr );
+#endif
```
Minor: Please add `/* MBEDTLS_X509_INFO */` here.

hanno-arm commented on library/oid.c:
```diff
@@ -46,6 +46,17 @@
+#else
+#define OID_DESCRIPTOR(s, name, description)  { ADD_LEN(s) }
+#define NULL_OID_DESCRIPTOR                   { NULL, 0 }
+#endif
```
Minor: Please add `/* MBEDTLS_X509_INFO */` here.

hanno-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -583,13 +586,19 @@ static int my_verify( void *data, mbedtls_x509_crt *crt,
+#else
+    ((void) crt);
+    ((void) depth);
+#endif
```
Minor: Please add `/* MBEDTLS_X509_INFO */` here.

hanno-arm commented:
@mpg Do you think you'll be able to find some time to re-review this, too?

hanno-arm commented:
@pkolbus Do you think you'll have time to finalize this PR, or shall we take over?

pkolbus commented:
Yes, I can continue working on this. Was waiting a bit for @mpg to respond and lost track of it.

pkolbus commented:
After rebasing, it looks like some/all of the `all.sh` tests are building with the crypto submodule, which is causing some build errors when `MBEDTLS_X509_INFO` is disabled. I think this design now requires a PR to mbed-crypto too, plus a new config for the `oid.c` and `oid.h` changes since x509 isn’t part of crypto. I can take care of all that.

Does this plan make sense? Would `MBEDTLS_OID_NAMES` be reasonable?

(If the mbedcrypto split is in a high degree of flux right now and it would make more sense to postpone this PR a bit, I’m happy to hold off)

Patater commented:
Hi @pkolbus,

Yes, the mbedcrypto split does complicate things a bit. It's not ideal right now, but because the previous libmbedcrypto library had oid.c in it, Mbed TLS currently gets that from Mbed Crypto. This is the final shape we want for these libraries, but the current shape is likely what we'll continue with for the next couple of months.

Could you raise any oid changes to Mbed Crypto? After that, make sure this PR includes an update to the submodule, pointing at a version of Mbed Crypto that has your oid changes.

Thanks, and sorry for the inconvenience.

hanno-arm commented:
Hi @pkolbus,

something I didn't spot earlier: The new compile-time option `MBEDTLS_X509_INFO` breaks the API for users which use a standalone configuration of Mbed TLS, because they'll loose the CRT info functionality unless they modify their configuration to include `MBEDTLS_X509_INFO`.

I'd suggest (ping @Patater) to negate the option to something like `MBEDTLS_X509_REMOVE_INFO` to resolve this.

Kind regards,
Hanno

hanno-arm commented:
Hi @pkolbus,

As you might be aware, your PR made it into the [baremetal](https://github.com/ARMmbed/mbedtls/blob/baremetal/include/mbedtls/config.h#L2044-L2053) branch, but unfortunately the development version got stuck.

Would you be willing to rebase the PR on top of development and negate the configuration option to `MBEDTLS_X509_REMOVE_INFO`, as explained in my last comment?

If you don't have time for this, I can also do the necessary rework - let me know what works best for you.

Thank you again for your contribution, it played an important in the RAM/ROM reductions that set the baremetal branch apart.

Kind regards,
Hanno

hanno-arm commented:
Superseded by https://github.com/ARMmbed/mbedtls/pull/3777 which is a rebased and reworked version of this PR.
